File: CodingSample/DotNet/Fundamentals/Functional/DelegateTest/DemoApp/Investment.cs
namespace DemoApp;

delegate double InterestRate(int years);

class Investment(double payment, int count)
{
    public double FutureValue(InterestRate scheme)
    {
        double i = scheme(count) / 100;
        return (payment / i) * (Math.Pow(1 + i, count) - 1);
    }
}

File: CodingSample/Java/Core/Functional/MethodRefTest/demo/Investment.java
//make static members of java.lang.Math accessible without class qualification
import static java.lang.Math.*;

class Investment {
    
    //a final instance field can only be assigned in a constructor
    private final double payment;
    private final int years;

    public Investment(double amount, int period) {
        payment = amount;
        years = period;
    }

    public double payment() {
        return payment;
    }

    public int years() {
        return years;
    }

    public double futureValue(InterestRate rate) {
        double i = rate.forPeriod(years) / 100;
        return (payment / i) * (pow(1 + i, years) - 1);
    }
}
